Added linux daemon

Chopikov opened the pull request:
- Added .cpp, .conf. .sh files
- Changed .gitignore

Chopikov committed:
Added linux daemon first version
- Added .cpp, .conf. .sh files
- Changed .gitignore

SWAGLORD1 commented:
При попытке собрать ваш код с помощью build.sh я получил несколько ошибок компиляции

KotsMikhail commented on surikov.ilya/lab1/daemon.cpp:
```diff
@@ -0,0 +1,187 @@
+        ipidFile >> prev;
+        sprintf(proc + 6, "%d", prev);
+        struct stat s;
+        if (!stat(proc, &s) && S_ISDIR(s.st_mode))
```
Зачем нужна эта проверка? Что пойдет не так, если послать SIGTERM несуществующему процессу?

KotsMikhail commented on surikov.ilya/lab1/daemon.cpp:
```diff
@@ -0,0 +1,187 @@
+
+
+void process_directory(const std::string& folder_src, const std::string& folder_dst, int mode) {
+    syslog(LOG_NOTICE, "Successfully working");
```
Это явно отладочная информация, которой в релизе быть не должно. Демон не должен спамить в лог каждый раз, когда он что-то делает

KotsMikhail commented on surikov.ilya/lab1/daemon.cpp:
```diff
@@ -0,0 +1,187 @@
+    }
+    cfg_path = argv[1];
+
+    kill_prev_daemon();
```
Прежде, чем убивать старого демона, стоит сначала успешно создать нового

KotsMikhail replied:
Прежде чем убивать старого демона, стоит сначала успешно создать нового

KotsMikhail commented on surikov.ilya/lab1/daemon.cpp:
```diff
@@ -0,0 +1,187 @@
+
+    umask(0);
+
+    sid = setsid();
```
Тут должно быть ещё одно ветвление, так как в некоторых системах лидер сессии автоматически получает управляющей терминал.

SWAGLORD1 replied:
Тут должно быть ещё одно ветвление, потому что в некоторых системах лидер сессии автоматически получает управляющий терминал

KotsMikhail commented on surikov.ilya/lab1/daemon.cpp:
```diff
@@ -0,0 +1,187 @@
+
+    if (argc != 2) {
+        printf("Wrong number of arguments. Expected: 2. Got: %d", argc);
+        return 0;
```
Тут должен быть exit, вы сейчас в дочернем процессе, у него нет своего терминала, этот код возврата никто не увидит

KotsMikhail commented on surikov.ilya/lab1/daemon.cpp:
```diff
@@ -0,0 +1,187 @@
+    signal(SIGHUP, signal_handler);
+    signal(SIGTERM, signal_handler);
+
+    set_pid_file();
```
Это, по идее, должно происходить вместе остановкой предыдущего демона, если он запущен

SWAGLORD1 replied:
Это, по идее, должно происходить одновременно с остановкой предыдущего демона, если он запущен

KotsMikhail commented on surikov.ilya/lab1/daemon.cpp:
```diff
@@ -0,0 +1,187 @@
+
+    set_pid_file();
+
+    while (1)
```
Используйте while(true), это более наглядный стиль кодирования

SWAGLORD1 replied:
Лучше использовать while(true), это более наглядный стиль кодирования

KotsMikhail commented on surikov.ilya/lab1/daemon.cpp:
```diff
@@ -0,0 +1,187 @@
+}
+
+void move_file(const std::string& path_src, const std::string& path_dst) {
+    system( (std::string("mv ") + path_src + std::string(" ") + path_dst).c_str() );
```
Из всех способов переместить файл вы выбрали самый ненадежный. Вызывать выполнение команды bash из кода С++ следует делать только в крайних случаях, потому что результат сильно зависит от платформы и её настроек. Лучше делать это средствами стандартных библиотек или библиотек Linux, есть много способов добиться нужного результата без bash команды

KotsMikhail commented on surikov.ilya/lab1/build.sh:
```diff
@@ -0,0 +1,12 @@
+
+chmod 666 "$pid_location"
+
+gcc -Wall -Werror daemon.cpp -o daemon_lab -lstdc++
```
Если вы используете 11 стандарт С++ то при сборке следует указывать флаг -std=c++11, без него на некоторых системах и версиях компилятора собираться не будет

KotsMikhail reviewed (changes requested):
В целом неплохо, ваша программа в целом правильно работает, но по коду есть некоторые замечания. Исправите - зачту.

Chopikov committed:
Fixed remarks
- Added flag to .sh file
- Fixed move files function
- Added second fork
- Refactored

KotsMikhail reviewed (changes requested):
Извиняюсь за ревью с опозданием, оповещение от гитхаба потерялось в куче писем
Теперь ваш демон, судя по логам, получает sigterm сразу после старта при запуске с любым конфигом.
Так же я заметил, что программа ничего не выдает при запуске без параметров, это тоже стоит исправить

Chopikov committed:
Fixed bug with process killing itself. Added writing info about wrong args numbers

KotsMikhail reviewed (approved):
Засчитано
